runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java

```diff
@@ -514,6 +528,27 @@ public synchronized void invalidateLayout() {
                             continue;
                         }
 
+                        // If this cluster was not what we expected
+                        if (l.getClusterId() == null // Legacy layout
+                                || !l.getClusterId().equals(clusterId)) {
+                            // We didn't have a clusterId set, so this is the
+                            // new clusterId. All future layouts must use this
+                            // clusterId.
+                            if (clusterId == null
+                                    // If the client requested a specific cluster, we
+                                    // check it here.
+                                    && (parameters.getClusterId() == null
+                                    || parameters.getClusterId().equals(l.getClusterId()))) {
+                                clusterId = l.getClusterId();
```

Comment:
if l.getCusterId == null and clusterId == null and parameters.getClusterId == null then the new clusterId will be null. Is that what we wanted ?
In general it is hard to read this code. It needs to be simplified.